### PR TITLE
[moe] refactor code to better adapt to llm

### PR DESCRIPTION
**Pull request by oahzxl**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

refactor code to better adapt to llm
1. remove abandoned usage like gpc
2. only use tp and ep ffn
3. add ckpt io for ffn
4. add more functions in api

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/nn/layer/moe/checkpoint.py [hunk 1]
```diff
@@ -15,26 +15,15 @@ def save_moe_model(model: nn.Module, save_path: str):
...
-                    load_param = state_dict[param_key]
-                    state_dict[cur_key] = load_param
-
-            for name, _ in module.experts[0].named_parameters():
-                pop_pre = f'{prefix}.experts.'
-                pop_suf = f'.{name}'
-                for i in range(num_local, module.num_total_experts):
-                    pop_key = f'{pop_pre}{i}{pop_suf}'
-                    state_dict.pop(pop_key)
+    for name, module in model.named_parameters():
```
Comment (ver217) [inline colossalai/nn/layer/moe/checkpoint.py:18]: Can you use checkpoint io or hook of state dict?

File: colossalai/nn/layer/moe/experts.py [hunk 2]
```diff
@@ -144,29 +84,57 @@ def forward(self, inputs):    # inputs [g, el, c, h]
         outputs = outputs.transpose(0, 1).contiguous()
         return outputs
 
+    def state_dict(self, destination=None, prefix='', keep_vars=False):
+        dp_rank = dist.get_rank(get_dp_group(self))
+        ep_rank = dist.get_rank(get_ep_group(self))
+        ep_size = get_ep_size(self)
+        # dp rank 0 will save the state dict
+        if dp_rank == 0:
+            for name, module in self.named_parameters():
```
Comment (ver217) [inline colossalai/nn/layer/moe/experts.py:93]: Is it a typo?

File: colossalai/tensor/moe_tensor/moe_info.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+from colossalai.tensor import ProcessGroup
```
Comment (ver217) [inline colossalai/tensor/moe_tensor/moe_info.py:1]: DON'T use this. This will be deprecated soon.

File: colossalai/nn/layer/moe/utils.py [hunk 4]
```diff
@@ -1,68 +1,70 @@
...
+
+    def __call__(self, inputs: torch.Tensor):
+        noisy = self.uniform(inputs.shape)
+        return inputs * noisy
+
+
+def autocast_softmax(logit: torch.Tensor, dim: int):
+    if logit.dtype != torch.float32:
+        logit = logit.float()
+    return F.softmax(logit, dim=dim)
```
Comment (ver217) [inline colossalai/nn/layer/moe/utils.py:60]: `torch.softmax()` receives `dtype` as arg

File: requirements/requirements.txt [hunk 5]
```diff
@@ -10,5 +10,5 @@ contexttimer
 ninja
 torch>=1.11
 safetensors
-flash_attn>=2.0
+flash-attn==2.0.5
```
Comment (ver217) [inline requirements/requirements.txt:13]: Remove this from requirements, as it's not essential.

File: colossalai/nn/layer/moe/experts.py (hunk 2 above)
Comment (oahzxl) [reply colossalai/nn/layer/moe/experts.py:97]: if you refer to ep, that means expert parallel. There is only tp process group, I use it to represent ep

File: requirements/requirements.txt (hunk 5 above)
Comment (oahzxl) [reply requirements/requirements.txt]: done

File: colossalai/nn/layer/moe/utils.py (hunk 4 above)
Comment (oahzxl) [reply colossalai/nn/layer/moe/utils.py]: done

File: colossalai/nn/layer/moe/checkpoint.py (hunk 1 above)
Comment (oahzxl) [reply colossalai/nn/layer/moe/checkpoint.py]: done

File: colossalai/tensor/moe_tensor/moe_info.py (hunk 3 above)
Comment (oahzxl) [reply colossalai/tensor/moe_tensor/moe_info.py]: update to ProcessGroupMesh

File: colossalai/nn/layer/moe/checkpoint.py [hunk 6]
```diff
@@ -1,40 +1,61 @@
...
+
+
+class MoeCheckpintIO(CheckpointIO):
+
+    def __init__(self) -> None:
+        super().__init__()
+
+    def load_unsharded_model(self, model: nn.Module, checkpoint: str, strict: bool):
+        state_dict = torch.load(checkpoint)
+        for name, module in model.named_parameters():
```
Comment (ver217) [inline colossalai/nn/layer/moe/checkpoint.py:20]: Should `module` be `param`?

File: colossalai/nn/layer/moe/checkpoint.py [hunk 7]
```diff
@@ -1,40 +1,61 @@
...
+                    else:
+                        state_dict.pop(new_name)
 
-from .experts import MoeExperts
+        model.load_state_dict(state_dict, strict=strict)
 
+    def save_unsharded_model(self, model: nn.Module, checkpoint: str, gather_dtensor: bool, use_safetensors: bool):
+        state_dict = model.state_dict()
+        if dist.get_rank() == 0:
+            torch.save(state_dict, checkpoint)
```
Comment (ver217) [inline colossalai/nn/layer/moe/checkpoint.py:36]: Does rank0 contain full state dict?

File: colossalai/nn/layer/moe/checkpoint.py (hunk 6 above)
Comment (oahzxl) [reply colossalai/nn/layer/moe/checkpoint.py:20]: done

File: colossalai/nn/layer/moe/checkpoint.py (hunk 7 above)
Comment (oahzxl) [reply colossalai/nn/layer/moe/checkpoint.py:36]: yes, i test it in test_moe_ckpt

File: colossalai/nn/layer/moe/experts.py (hunk 2 above)
Comment (oahzxl) [reply colossalai/nn/layer/moe/experts.py]: done

File: colossalai/nn/layer/moe/experts.py [hunk 8]
```diff
@@ -1,204 +1,136 @@
...
+                 num_experts: int,
+                 hidden_size: int,
+                 intermediate_size: int,
+                 activation=None,
+                 drop_rate: float = 0):
+        super().__init__(num_experts, hidden_size, intermediate_size, "EP", activation, drop_rate)
 
-        e = inputs.size(1)
-        h = inputs.size(-1)
+    def state_dict(self, destination=None, prefix='', keep_vars=False):
```
Comment (ver217) [inline colossalai/nn/layer/moe/experts.py:91]: Does `keep_vars` affect this method?

Comment (oahzxl) [reply colossalai/nn/layer/moe/experts.py:91]: `keep_vars` is a required arg in state dict, so i add it. Now it does affect

**Review by ver217 — APPROVED** [review]
